File: advance/1039. Course List for Student (25).cpp
//-------------------------------最后一个测试点超时 
#include<iostream>
#include<map>
#include<set>
#include<string>
using namespace std;
int n;//学生数量
int k;//课程数量
int main() {
	string stu;
	int index,num;
	cin >> n >> k;
	map<string, set<int>> ma;
	for (int i = 0; i < k; ++i) {
		cin >> index>>num;
		for (int j = 0; j < num; ++j) {
			cin >> stu;
			ma[stu].insert(index);
		}
	}
	for (int i = 0; i < n; ++i) {
		cin >> stu;
		cout << stu;
		cout << " " << ma[stu].size();
		if (ma.find(stu) != ma.end()) {
			for (set<int>::iterator it = ma[stu].begin(); it != ma[stu].end(); ++it)
				cout << " " << *it;
		}
		cout << endl;
	}
	return 0;
}

File: advance/java/1039. Course List for Student (25).java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Scanner;
import java.util.Set;

public class Main{
	public static void main(String[] args) {
		Scanner in=new Scanner(System.in);
		int n=in.nextInt(),k=in.nextInt();
		Map<String,Set<Integer>> ma=new HashMap<String,Set<Integer>>(); 
		for(int i=0;i<k;++i) {
			int index=in.nextInt();
			int num=in.nextInt();
			for(int j=0;j<num;++j) {
				String stu=in.next();
				Set<Integer> s=new HashSet<>();
				if(ma.get(stu)!=null)s=ma.get(stu);
				s.add(index);
				ma.put(stu, s);
			}
		}
		for(int i=0;i<n;++i) {
			String stu=in.next();
			System.out.print(stu+" ");
			if(ma.get(stu)!=null) {
				System.out.print(ma.get(stu).size());
				for(Integer str:ma.get(stu))System.out.print(" "+str);
			}else System.out.print("0");
			System.out.println();
		}
	}

}